sql/parser/eval.go

```diff
@@ -795,7 +795,11 @@ func evalFuncExpr(expr *FuncExpr) (Datum, error) {
 		types = append(types, reflect.TypeOf(arg))
 	}
 
-	if expr.fn.fn == nil {
+	// Aggregate functions get looked up fresh every time. This is necessary
+	// because the argument type can change between when the first lookup during
+	// normalization and when the function is executed. For example, COUNT(*)
+	// will lookup as COUNT(tuple) but then get executed as COUNT(int).
+	if expr.fn.fn == nil || expr.fn.aggregate {
```

Comment:
is there any value in modeling the aggregate functions as builtins like this? seems like we could return the first argument directly here after checking for present in the `aggregates` map.